crates/runtime/src/component.rs

```diff
@@ -0,0 +1,431 @@
+            }
+            for i in 0..self.offsets.num_runtime_reallocs {
+                let i = RuntimeReallocIndex::from_u32(i);
+                let offset = self.offsets.runtime_realloc(i);
+                *self.vmctx_plus_offset(offset) = INVALID_PTR;
+            }
+        }
+    }
+}
+
+impl VMComponentContext {
+    /// Moves the `self` pointer backwards to the `ComponentInstance` pointer
+    /// that this `VMComponentContext` trails.
+    pub fn instance(&self) -> *mut ComponentInstance {
+        unsafe {
+            (self as *const Self as *mut u8)
+                .offset(-(offset_of!(ComponentInstance, vmctx) as isize))
+                as *mut ComponentInstance
+        }
+    }
+}
+
+/// An owned version of `ComponentInstance` which is akin to
+/// `Box<ComponentInstance>`.
+///
+/// This type can be dereferenced to `ComponentInstance` to access the
+/// underlying methods.
+pub struct OwnedComponentInstance {
+    ptr: ptr::NonNull<ComponentInstance>,
+}
+
+// Using `NonNull` turns off auto-derivation of these traits but the owned usage
+// here enables these trait impls so long as `ComponentInstance` itself
+// implements these traits.
+unsafe impl Send for OwnedComponentInstance where ComponentInstance: Send {}
+unsafe impl Sync for OwnedComponentInstance where ComponentInstance: Sync {}
+
+impl OwnedComponentInstance {
+    /// Allocates a new `ComponentInstance + VMComponentContext` pair on the
+    /// heap with `malloc` and configures it for the `component` specified.
+    pub fn new(component: &Component, store: *mut dyn Store) -> OwnedComponentInstance {
+        let offsets = VMComponentOffsets::new(HostPtr, component);
+        let layout = ComponentInstance::alloc_layout(&offsets);
+        unsafe {
+            // Technically it is not required to `alloc_zeroed` here. The
+            // primary reason for doing this is because a component context
+            // start is a "partly initialized" state where pointers and such are
+            // configured as the instantiation process continues. The component
+            // model should guarantee that we never access uninitialized memory
+            // in the context, but to help protect against possible bugs a
+            // zeroed allocation is done here to try to contain
+            // use-before-initialized issues.
+            let ptr = alloc::alloc_zeroed(layout) as *mut ComponentInstance;
+            let ptr = ptr::NonNull::new(ptr).unwrap();
+
+            ComponentInstance::new_at(ptr.as_ptr(), layout.size(), offsets, store);
+
+            OwnedComponentInstance { ptr }
+        }
+    }
+}
+
+impl Deref for OwnedComponentInstance {
+    type Target = ComponentInstance;
+    fn deref(&self) -> &ComponentInstance {
+        unsafe { &*self.ptr.as_ptr() }
+    }
+}
+
+impl DerefMut for OwnedComponentInstance {
+    fn deref_mut(&mut self) -> &mut ComponentInstance {
+        unsafe { &mut *self.ptr.as_ptr() }
+    }
+}
```

Comment:
This isn't technically safe right? Since the `ComponentInstance` is logically pinned, giving out an `&mut` would allow moving it out by `std::mem::replace`ing it.

Reply:
That's a great point actually and something I totally missed! I removed this, made it a private internal accessor, and reexported `&mut` self methods from the un-owned version to the owned version.